### DISCUSSION TITLE
如何让模型理解工具返回的数据？

### DISCUSSION THREAD

**Discussion by minmie** [zai-org/ChatGLM3#502] [Q&A] [answered]
官方提供的查询天气demo，工具返回了如下json
```
{'current_condition': {'temp_C': '3', 'FeelsLikeC': '-0', 'humidity': '87', 'weatherDesc': [{'value': 'Overcast'}], 'observation_time': '09:43 AM'}}
```
模型可以根据这个结果做出正确的回答。

1.请问模型是怎么做到的？我看prompt只有对工具的传参做说明，并没有对返回做说明

2.我用自定义的工具返回json，模型都不能理解，这又是为什么？

**Comment by zRzRzRzRzRzRzR** [comment]
是直接理解，你的返回就是提示词拼接之前你提问的历史让模型理解的，所以如果你自定义返回的json过于复杂，或者没有在相关提示词内告诉模型返回的是什么，那么模型是理解不了的

**Reply by minmie** [reply]
@zRzRzRzRzRzRzR 
1. 你的原话“或者没有在相关提示词内告诉模型返回的是什么，那么模型是理解不了的”
但是官方提供的天气demo，也没有对天气工具的返回的json做说明啊，只对工具的入参做了说明，但是模型依然能理解，这是为什么？

2. 如果我需要让模型理解返回的json,我需要怎么做？这部分的提示词在哪里写？

**Reply by zRzRzRzRzRzRzR** [reply]
1. 官方dmo 中返回的内容极其简单 如果你打开那个json发现只有一个两个内容，在我们的demo中，我们提问的是 温度是多少
json返回如下
{'temperature': '10', 'description': '晴'}
也就是 我问温度是多少，json返回的内容就是 温度是多少。基本没有理解难度。经过实际测试，如果返回内容多，需要理解和转义是有一定难度的
2. 关于你的第二个问题，你需要微调整套工具调用，在observation之后，有一个thought，这个地方是你直接写对参数理解的内容

**Reply by minmie** [reply]
![image](https://github.com/THUDM/ChatGLM3/assets/40080081/5ce71b00-b00a-4a00-aa50-0be36c8bead4)
1.你好，对于你说的“在observation之后，有一个thought”。我看了官方的说明，只有在tool调用之前有一个thought，在observation后并没有thought。或者你的意思是我应该在observation之后自己加一个thought?

2. 另外，这个thought应该如何去写呢？假设说工具返回的observation为:
{'A': '老王', 'B': '男', ‘C’：28}
thought像下面这样的写法对吗？
“A代表姓名，B代表性别，C代表年龄”
（这样写感觉也有点奇怪，thought也是模型自己生成的，既然模型能生成这样的thought，表明模型本身就已经理解了这些字段的含义，那还要写这个thought是不是有点多此一举了？）

@zRzRzRzRzRzRzR

**Reply by zRzRzRzRzRzRzR** [reply, marked as answer]
1. 是的 你看格式的是 response to observation
2. 不是，应该写， 我得到了信息，用户查询的人是 老王 性别是男，28岁
